src/wallet/sqlite.cpp

```diff
@@ -42,6 +43,18 @@ SQLiteDatabase::SQLiteDatabase(const fs::path& dir_path, const fs::path& file_pa
     }
 }
 
+bool SQLiteDatabase::PrepareDirectory() const
+{
+    if (m_mock) return true;
+    // Try to create the directory containing the wallet file and lock it
+    TryCreateDirectories(m_dir_path);
+    if (!LockDirectory(m_dir_path, ".walletlock")) {
```

Comment:
In commit "sqlitedb: Create and lock the wallet directory" (af798a8fd9b3b1a1e0a63a665f6d89c73e8614a3)

This doesn't seem like the most ideal approach. We were forced to use .walletlock in #11904 because the old version of BDB we use doesn't support set_lk_exclusive. But with sqlite it would seem simpler and safer to use its exclusive locking functionality: https://www.sqlite.org/pragma.html#pragma_locking_mode instead of rolling our own which is probably [less portable and reliable](https://github.com/bitcoin/bitcoin/issues/19167#issuecomment-698594254), requires code duplication, and is nonstandard not working with other sqlite database tools.

Reply:
Since only an EXCLUSIVE lock is acquired after the first write, I'm not sure that the locking pragma provides strong enough guarantees for us. Another bitcoind could conceivably open the same wallet while only the SHARED lock is held.

Reply:
From https://sqlite.org/pragma.html#pragma_locking_mode

> The first time the database is read in EXCLUSIVE mode, a shared lock is obtained and held. The first time the database is written, an exclusive lock is obtained and held.

So opening in exclusive mode means that there is a potential case where bitcoind has not yet written anything and thus only has a shared lock. In that case, another bitcoind could open the same file (and not write anything) and also acquire a shared lock. See https://sqlite.org/lockingv3.html for the definitions of shared and exclusive locks in sqlite3.

Reply:
> This just means you need 3 statements to open the database:

Then there's the potential for race conditions in 2 instances: before the pragma is executed, and after it is executed but before the transaction begins.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/19077#discussion_r501156002

> Then there's the potential for race conditions in 2 instances: before the pragma is executed, and after it is executed but before the transaction begins.

There's no race condition here:

```c++
db = sqlite_open(path); // Or error "sorry, couldn't open database"
sqlite_exec(db, "pragma locking_mode=exclusive"); // Or error "sorry, couldn't set locking mode"
sqlite_exec(db, "begin exclusive transaction"); // Or error "sorry, database is currently in use"
sqlite_exec(db, "commit"); // Or error "sorry, unexpected failure"
```

There would be a race condition if the exclusive lock was released after the commit statement. But the point of the pragma is to extend the exclusive lock from point where it is first acquired until the point where the database is closed.

Reply:
Done